core/src/oracle_state.rs

```diff
@@ -94,14 +97,33 @@ pub struct Stage {
 #[derive(Debug)]
 pub struct OraclePool {
     pub data_point_source: Box<dyn DataPointSource>,
-    /// Stages
-    pub datapoint_stage: Stage,
-    // Local Oracle Datapoint Scan
-    pub local_oracle_datapoint_scan: Option<Scan>,
+    /// Stages. Note it's defined as a tuple with `OracleContractParameters` and `TokenIds`. We need
+    /// to do this because the implementation of the `DatapointBoxesSource` trait requires each of
+    /// these tuple elements.  Now our codebase makes use of `&dyn DatapointBoxesSource`, and trying
+    /// to use an instance of `OracleContractParameters` within the implementation of the trait
+    /// leads to trouble with Rust's borrow checker.
+    ///
+    /// More specifically, using a separate `OracleContractParameters` instance within the trait
+    /// implementation results in the creation of a local variable whose reference is needed but
+    /// is immediately dropped. Type `rustc --explain E0515` on the command line for details.
+    ///
+    /// Note: need to use `Arc` here since it will be passed between threads due to the `api`
+    /// module.
+    pub datapoint_stage: (Stage, Arc<OracleContractParameters>, Arc<TokenIds>),
```

Comment:
Could you consider making structs instead of using a tuple? I mean, three elements in a tuple is a sign. :) Plus, it could use only needed token ids.

Reply:
Yeah it is indeed a sign : ) I'll fix it.